### ISSUE TITLE
Error recovery can trigger Assertion `DefinitionData && "queried property of class with no definition"' (DeclCXX.h:617)

### ISSUE THREAD

**Issue by hfinkel** [llvm/llvm-project#13149]
|  |  |
| --- | --- |
| Bugzilla Link | [12777](https://llvm.org/bz12777) |
| Version | trunk |
| OS | Linux |
| Attachments | [reduced test case (provided by C.S.)](https://user-images.githubusercontent.com/910929/143745268-ac2bc125-765e-4550-b892-e9733c9c0bc1.gz) |
| CC | @DougGregor,@rbock |

## Extended Description 

Running clang trunk on x86_64 on:
```cpp
 template<class _CharT> struct char_traits;
     template<typename _CharT, typename _Traits = char_traits<_CharT> > class basic_ios;
   class ios_base {
      iostate _M_exception;
    template<typename _CharT, typename _Traits> basic_ios<_CharT, _Traits>& basic_ios<_CharT, _Traits>::copyfmt(const basic_ios& __rhs) {
        _M_exception = goodbit;
```

results in:

In file included from prog.cpp:1:
prog.cpp:4:7: error: unknown type name 'iostate'
      iostate _M_exception;
      ^
prog.cpp:5:119: error: use of class template basic_ios requires template arguments
    template<typename _CharT, typename _Traits> basic_ios<_CharT, _Traits>& basic_ios<_CharT, _Traits>::copyfmt(const basic_ios& __rhs) {
                                                                                                                      ^
prog.cpp:2:79: note: template is declared here
     template<typename _CharT, typename _Traits = char_traits<_CharT> > class basic_ios;
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~       ^
prog.cpp:5:105: error: out-of-line definition of 'copyfmt' from class 'basic_ios<_CharT, _Traits>' without definition
    template<typename _CharT, typename _Traits> basic_ios<_CharT, _Traits>& basic_ios<_CharT, _Traits>::copyfmt(const basic_ios& __rhs) {
                                                                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
prog.cpp:6:32: error: expected '}'
        _M_exception = goodbit;
                               ^
prog.cpp:3:19: note: to match this '{'
   class ios_base {
                  ^
clang: /src/llvm-trunk-writable/tools/clang/lib/AST/../../include/clang/AST/DeclCXX.h:617: const clang::CXXRecordDecl::DefinitionData& clang::CXXRecordDecl::data() const: Assertion `DefinitionData && "queried property of class with no definition"' failed.

**Comment by shafik** [comment]
Does not reproduce on trunk: https://godbolt.org/z/bzafbee19

**Comment by Endilll** [comment]
Both the first and the third examples were fixed in Clang 3.7:
https://godbolt.org/z/3c9Wf6aTW
https://godbolt.org/z/8vo6dzfv5
